src/statsdproto.rs

```diff
@@ -0,0 +1,385 @@
+                Err(_) => Err(ParseError::InvalidSampleRate),
+            })
+            .transpose()?;
+        let mtype: Type = pdu.pdu_type().try_into()?;
+        let tags = pdu.tags().map(|v| parse_tags(v)).transpose()?;
+        Ok(Parsed {
+            name: pdu.name().to_vec(),
+            mtype: mtype,
+            value: value,
+            sample_rate: sample_rate,
+            tags: tags.unwrap_or_default(),
+        })
+    }
+}
+
+/// A PDU is an incoming protocol unit for statsd messages, commonly a
+/// single datagram or a line-delimitated message. This PDU type owns an
+/// incoming message and can offer references to protocol fields. It only
+/// performs limited parsing of the protocol unit.
+#[derive(Debug, Clone)]
+pub struct PDU {
+    underlying: Bytes,
+    value_index: usize,
+    type_index: usize,
+    type_index_end: usize,
+    sample_rate_index: Option<(usize, usize)>,
+    tags_index: Option<(usize, usize)>,
+}
+
+impl PDU {
+    pub fn name(&self) -> &[u8] {
+        &self.underlying[0..self.value_index - 1]
+    }
+
+    pub fn value(&self) -> &[u8] {
+        &self.underlying[self.value_index..self.type_index - 1]
+    }
+
+    pub fn pdu_type(&self) -> &[u8] {
+        &self.underlying[self.type_index..self.type_index_end]
+    }
+
+    pub fn tags(&self) -> Option<&[u8]> {
+        self.tags_index.map(|v| &self.underlying[v.0..v.1])
+    }
+
+    pub fn sample_rate(&self) -> Option<&[u8]> {
+        self.sample_rate_index.map(|v| &self.underlying[v.0..v.1])
+    }
+
+    pub fn len(&self) -> usize {
+        self.underlying.len()
+    }
+
+    pub fn as_ref(&self) -> &[u8] {
+        self.underlying.as_ref()
+    }
+
+    ///
+    /// Return a clone of the PDU with a prefix and suffix attached to the statsd name
+    ///
+    pub fn with_prefix_suffix(&self, prefix: &[u8], suffix: &[u8]) -> Self {
+        let offset = suffix.len() + prefix.len();
+
+        let mut buf = bytes::BytesMut::with_capacity(self.len() + offset);
+        buf.put(prefix);
+        buf.put(self.name());
+        buf.put(suffix);
+        buf.put(self.underlying[self.value_index - 1..].as_ref());
+
+        PDU {
+            underlying: buf.freeze(),
+            value_index: self.value_index + offset,
+            type_index: self.type_index + offset,
+            type_index_end: self.type_index_end + offset,
+            sample_rate_index: self
+                .sample_rate_index
+                .map(|(b, e)| (b + offset, e + offset)),
+            tags_index: self.tags_index.map(|(b, e)| (b + offset, e + offset)),
+        }
+    }
+
+    /// Parse an incoming single protocol unit and capture internal field
+    /// offsets for the positions and lengths of various protocol fields for
+    /// later access. No parsing or validation of values is done, so at a low
+    /// level this can be used to pass through unknown types and protocols.
+    pub fn new(line: Bytes) -> Option<Self> {
```

Comment:
I was caught a little off-guard by this signature initially — it's not super common to see `new` return an Option vs a Self or a Result<Self> IME.

I wonder if there'd be value in writing a TagParseFailure enum to express which failure mode caused construction to abort, since there's a couple ways this can `None` out, if for no other reason than it makes identifying the correct failure mode easier from tests. I suppose it'd also make it easier to propagate the exact error out via ParseError too.

Reply:
Yeah, ParseError came into existence after this was written. Will take a note to refactor this with more specific error output.